src/Network/GRPC/MQTT/Sequenced.hs

```diff
@@ -104,11 +108,32 @@ mkSequencedRead read = do
 
   return $ atomically orderedRead
 
-mkPacketizedPublish :: forall io io'. (MonadIO io, MonadIO io') => MQTTClient -> Int64 -> Topic -> io (LByteString -> io' ())
+mkStreamRead :: forall r io. (Message r, MonadIO io) => ExceptT RemoteError IO LByteString -> io (ExceptT RemoteError IO (Maybe r))
+mkStreamRead readRequest = do
+  reqsRef :: IORef (Maybe (Vector r)) <- newIORef (Just Vector.empty)
+
+  let readNextChunk :: ExceptT RemoteError IO ()
+      readNextChunk = do
+        res <- readRequest
+        cs <- hoistEither $ unwrapStreamChunk res
+        writeIORef reqsRef cs
+
+  let readStreamChunk :: ExceptT RemoteError IO (Maybe r)
+      readStreamChunk =
+        readIORef reqsRef >>= \case
+          Nothing -> pure Nothing
+          Just reqs ->
+            if Vector.null reqs
+              then readNextChunk >> readStreamChunk
+              else writeIORef reqsRef (Just $ Vector.tail reqs) >> return (Just $ Vector.head reqs)
```

Comment:
After only one "chunk" this will always push the tail back into the IORef and then return the head. Why not just discard the tail since you're no longer going to use it? Unless I'm missing something more subtle...

Reply:
`readStreamChunk` gets called repeatedly to retrieve values of type `Just r` till it runs out of data and eventually returns `Nothing`. So we need to keep values in the tail.

Reply:
Oh I see, the expression `readStreamChunk` is called in the StreamRecv of the GRPC stream handler then?

Reply:
Correct. `mkStreamRead` is passed to GRPC stream handler functions and gets invoked repeatedly.